### PR TITLE
Tkurth/rl tests

### PR DESCRIPTION
**Pull request by azrael417**
This PR fixes a bug in SAC, where instead of alpha was read from the file and used as entropy coefficient, rho was used instead which made training unstable since alpha should be around 0.1 and rho is close to 1.0. In addition to this fix, this PR adds trainable entropy coefficient as it can improve stability of training.

Furthermore, this PR adds a test suite for many parts of the RL stack and in addition a full test for each algorithm (DDPG, TD3, SAC, PPO) on simple tests environments. Those tests run for some time but it is helpful to verify that the environment works. The only test which is currently failing is the one for the state dependent action reward for DDPG, and only for DDPG. This could have something to do with the fact that DDPG sometimes overfits rapidly on simple environments, so I would not assume that the DDPG implementation by itself is wrong. In any case, since TD3 is basically always superior to DDPG and it seems to work in all cases, I recommend using that instead.

In order for the tests to work, I included a gtest dependency and also added some simple RL related models as base models to torch fort. Those include a sac-policy and an actor-critic model, where actor and critic share the same feature extractor.

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -70,7 +70,8 @@ RUN cd /torchfort && mkdir build && cd build && \
     cmake -DCMAKE_INSTALL_PREFIX=/opt/torchfort \
     -DTORCHFORT_YAML_CPP_ROOT=/opt/yaml-cpp \
     -DTORCHFORT_NCCL_ROOT=/opt/nccl/build \
-    -DTORCHFORT_BUILD_EXAMPLES=1 \
+    -DTORCHFORT_BUILD_EXAMPLES=0 \
```
Comment (romerojosh) [inline docker/Dockerfile:73]: Did you intend to disable building examples?

File: src/csrc/include/internal/rl/rollout_buffer.h [hunk 2]
```diff
@@ -151,25 +173,35 @@ class GAELambdaRolloutBuffer : public RolloutBuffer, public std::enable_shared_f
...
-    auto ret_list = std::vector<float>(batch_size);
-
-    // we need to exclude the last element here, since it is a closed interval
-    std::uniform_int_distribution<size_t> uniform_dist(0, size_ - 1);
-    for (int sample = 0; sample < batch_size; sample++) {
-      auto index = uniform_dist(rng_);
+    std::vector<torch::Tensor> stens_list, atens_list;
+    std::vector<float> q_list, log_p_list, adv_list, ret_list;
+
+    //std::uniform_int_distribution<size_t> uniform_dist(0, size_ - 1);
```
Comment (romerojosh) [inline src/csrc/include/internal/rl/rollout_buffer.h:179]: Remove.

File: src/csrc/include/internal/rl/rollout_buffer.h [hunk 3]
```diff
@@ -151,25 +173,35 @@ class GAELambdaRolloutBuffer : public RolloutBuffer, public std::enable_shared_f
...
+    // go through the buffer in random order
+    auto lower = pos_;
+    auto upper = std::min(pos_ + static_cast<size_t>(batch_size), size_);
+    for (size_t sample = lower; sample < upper; sample++) {
+      //auto index = uniform_dist(rng_);
```
Comment (romerojosh) [inline src/csrc/include/internal/rl/rollout_buffer.h:184]: Remove.

File: src/csrc/rl/off_policy/sac.cpp [hunk 4]
```diff
@@ -179,6 +218,15 @@ SACSystem::SACSystem(const char* name, const YAML::Node& system_node,
     THROW_INVALID_USAGE("Missing critic_lr_scheduler block in configuration file.");
   }
 
+  // alpha parameter
+#if 0
```
Comment (romerojosh) [inline src/csrc/rl/off_policy/sac.cpp:222]: Remove unused code block here if not needed.

File: src/csrc/rl/off_policy/sac.cpp [hunk 5]
```diff
@@ -393,10 +503,10 @@ void SACSystem::loadCheckpoint(const std::string& checkpoint_dir) {
 // we should pass a tuple (s, a, s', r, d)
 void SACSystem::updateReplayBuffer(torch::Tensor s, torch::Tensor a, torch::Tensor sp, float r, bool d) {
   // note that we have to rescale the action: [a_low, a_high] -> [-1, 1]
-  auto as = scale_action(a, a_low_, a_high_);
+  //auto as = scale_action(a, a_low_, a_high_);
```
Comment (romerojosh) [inline src/csrc/rl/off_policy/sac.cpp:506]: Remove.

**Review by romerojosh — COMMENTED** [review]
Added mostly nitpick comments here. Otherwise changes LGTM. Thanks for adding some tests!

File: docker/Dockerfile (hunk 1 above)
Comment (azrael417) [reply docker/Dockerfile]: I am sorry, I think I disabled that during the debug phase. I have enabled it again.

File: src/csrc/include/internal/rl/rollout_buffer.h (hunk 2 above)
Comment (azrael417) [reply src/csrc/include/internal/rl/rollout_buffer.h]: fixed.

File: src/csrc/rl/off_policy/sac.cpp (hunk 4 above)
Comment (azrael417) [reply src/csrc/rl/off_policy/sac.cpp]: That one is actually needed, I removed the comment.

File: src/csrc/rl/off_policy/sac.cpp (hunk 5 above)
Comment (azrael417) [reply src/csrc/rl/off_policy/sac.cpp]: fixed.

File: src/csrc/include/internal/rl/rollout_buffer.h (hunk 3 above)
Comment (azrael417) [reply src/csrc/include/internal/rl/rollout_buffer.h]: fixed.

**Comment by azrael417** [comment]
I have addressed all your comments and agree to all of them, thanks for the careful review.

**Review by romerojosh — APPROVED** [review]
LGTM!
